pkg/trojan/message.go

```diff
@@ -0,0 +1,283 @@
+	if len(data) < 74 {
+		return false
+	}
+
+	// check for valid trojan message length in bytes #41 and #42
+	messageLen := int(binary.BigEndian.Uint16(data[40:42]))
+	return 74+messageLen <= len(data)
+}
+
+// checkTargets verifies that the list of given targets is non empty and with elements of matching size
+func checkTargets(targets Targets) error {
+	if len(targets) == 0 {
+		return ErrEmptyTargets
+	}
+	validLen := len(targets[0]) // take first element as allowed length
+	for i := 1; i < len(targets); i++ {
+		if len(targets[i]) != validLen {
+			return ErrVarLenTargets
+		}
+	}
+	return nil
+}
+
+// toChunk finds a nonce so that when the given trojan chunk fields are hashed, the result will fall in the neighbourhood of one of the given targets
+// this is done by iteratively enumerating different nonces until the BMT hash of the serialization of the trojan chunk fields results in a chunk address that has one of the targets as its prefix
+// the function returns a new chunk, with the found matching hash to be used as its address,
+// and its data set to the serialization of the trojan chunk fields which correctly hash into the matching address
+func (m *Message) toChunk(targets Targets, span []byte) (swarm.Chunk, error) {
+	// start out with random nonce
+	nonce := make([]byte, 32)
+	if _, err := rand.Read(nonce); err != nil {
+		return nil, err
+	}
+	nonceInt := new(big.Int).SetBytes(nonce)
+	targetsLen := len(targets[0])
+
+	// serialize message
+	b, err := m.MarshalBinary() // TODO: this should be encrypted
+	if err != nil {
+		return nil, err
+	}
+
+	// hash chunk fields with different nonces until an acceptable one is found
+	// TODO: prevent infinite loop
+	for {
+		s := append(append(span, nonce...), b...) // serialize chunk fields
+		hash1, err := hashBytes(s)
+		if err != nil {
+			return nil, err
+		}
+
+		// take as much of the hash as the targets are long
+		if contains(targets, hash1[:targetsLen]) {
+			// if nonce found, stop loop and return chunk
+			return swarm.NewChunk(swarm.NewAddress(hash1), s), nil
+		}
+		// else, add 1 to nonce and try again
+		nonceInt.Add(nonceInt, big.NewInt(1))
+		// loop around in case of overflow
+		if nonceInt.BitLen() > 256 {
+			nonceInt = big.NewInt(0)
+		}
+		nonce = padBytes(nonceInt.Bytes()) // pad in case Bytes call is not 32 bytes long
+	}
+}
+
+// hashBytes hashes the given serialization of chunk fields with the hashing func
+func hashBytes(s []byte) ([]byte, error) {
+	if hasher == nil {
+		return nil, ErrInvalidHasher
+	}
+	hasher.Reset()
+	span := binary.LittleEndian.Uint64(s[:8])
+	err := hasher.SetSpan(int64(span))
+	if err != nil {
+		return nil, err
+	}
+	if _, err := hasher.Write(s[8:]); err != nil {
+		return nil, err
+	}
+	return hasher.Sum(nil), nil
+}
+
+// contains returns whether the given collection contains the given elem
+func contains(col Targets, elem []byte) bool {
+	for i := range col {
+		if bytes.Equal(elem, col[i]) {
```

Comment:
This is byte-level matching, not bit level. Is that correct?

Would looking at the "proximity" functions in `pkg/swarm` be relevant here?

Reply:
correct. i am too ignorant to know whether `Proximity` is faster than `bytes.Equal`, but this should be replaced by bit-level matching some point

Reply:
at. what point? :0

Reply:
i suppose at the point that the `Targets` become bit vectors

Reply:
I would say that should be done now... :)

Reply:
i disagree. I would put bitvectors implementation in another PR (#443)  as an improvement. otherwise we would be stuck in this PR for ever.

Reply:
Ok then I would seriously encourage that this is the next thing that should happen.